# Nama: Syauqi Nabil Tasri

# NRP: 5054241040

<center><strong>Natural Language Processing</strong><br />
<strong><font color="blue">Semester Genap ITS</font></strong><br />
</center>

<strong>Outline pertemuan minggu ke-3</strong><br />
<li> Membaca dokumen dari file </li>
<li> Representasi vektor </li>
<li> Pembobotan</li>


### Membaca dokumen dari file
<p> Untuk membaca dokumen dari file dapat dilakukan dengan berbagai macam cara, bergantung pada format dari dokumen tersebut </p>

<p><img alt="" src="figures/2_importing_Files.jpg" style="height:400px; width:510px" /><br />
[<a href="http://www.clearcounsel.com/what-happens-to-your-digital-property-after-you-die/" target="_blank"><strong>Image Source</strong></a>]</p>

<p> Berikut ini merupakan list library untuk memparsing dokumen dalam berbagai format </p>
<p>* <a href="https://textract.readthedocs.io/en/stable/&amp;nbsp" target="_blank">https://textract.readthedocs.io/</a> <img alt="" src="figures/2_textract.png" style="height: 239px; width: 600px;" /></p>


In [2]:
# Membaca dokumen dengan format .txt
import os

src_name = "the_matrix_synopsis.txt"
path = "data"
src_path = os.path.join(path, src_name)

docs = list()
with open(src_path, 'r', encoding="utf-8") as fin:
    # membaca seluruh isi dari dokumen
    data = fin.readlines()

    # menghilangkan newline ("\n") dan baris yang kosong
    for d in data:
        d_strip = d.strip()

        if d_strip:
            docs.append(d_strip)

# New Section

In [3]:
# Mengecek banyaknya dokumen (note: 1 dokumen = 1 baris)
print("Panjang dokumen: ", len(docs))

# Mengakses isi dari dokumen dengan indeks 1
print(docs[1])

Panjang dokumen:  97
A phone rings and text appears on the screen: "Call trans opt: received. 2-19-98 13:24:18 REC: Log>" As a conversation takes place between Trinity (Carrie-Anne Moss) and Cypher (Joe Pantoliano), two free humans, a table of random green numbers are being scanned and individual numbers selected, creating a series of digits not unlike an ordinary phone number, as if a code is being deciphered or a call is being traced.


In [ ]:
# Membaca dokumen dengan format .csv
import os
import pandas as pd

src_name = "data_tweet.csv"
path = "Word Embedding/data"
src_path = os.path.join(path, src_name)

dff = pd.read_csv(src_path)

print(dff)

           Date       Screen Name                          Full Name  \
0     6/27/2018     @BachtiarLiem                         Webindo.id   
1     6/27/2018  @wikipediaberita                       #BERITAPedia   
2     6/27/2018     @kompaskampus                      Kompas Kampus   
3     6/27/2018     @GidiyonDavid                      Gidiyon David   
4     6/27/2018  @2019Ada_PilPres                            Yisreal   
...         ...               ...                                ...   
2419  6/27/2018      @ahmadvision                        Ahmad Ahmad   
2420  6/27/2018  @HorasSumutNews8                   Horas Sumut News   
2421  6/27/2018      @serambinews                  Serambi Indonesia   
2422  6/27/2018      @inibukanedo                    Edward Fernando   
2423  6/27/2018      @tareqAzizii  SUARA RAKYAT MISKIN JAKARTA(SRMJ)   

                                             Tweet Text      Tweet ID  \
0     Pilkada 2018, PDIP Menang di 6 Pilgub dan 91 D...  1.012

<h3 id="Json-Files">Json Files</h3>

<ul>
	<li>Populer digunakan untuk data dari Media Sosial dan NoSQL</li>
	<li>Portable: File Json memuat nama variabel dan nilainya (tidak seperti XML)</li>
	<li>Plain Text</li>
	<li>Schemaless: Setiap record tidak harus memiliki jumlah field yang tetap seperti csv</li>
	<li>JSON isomorfis dengan &quot;Dictionary&quot; di Python</li>
	<li>Contoh struktur file json:</li>
</ul>

<p><img alt="" src="figures/json.png" style="width: 200px; height: 211px;" /></p>


In [4]:
# Membaca dokumen dengan format .json
import json

src_name = "contoh.json"
path = "data"
src_path = os.path.join(path, src_name)

with open(src_path, 'r', encoding="utf-8", errors='ignore') as fin:
    data = json.load(fin)

data

{'nama': 'Mukidi', 'umur': 19, 'mobil': 'BMW', 'motor': 'Forza'}

In [6]:
pip install bs4

  Using cached typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)
   ---------------------------------------- 0.0/107.7 kB ? eta -:--:--
   --- ------------------------------------ 10.2/107.7 kB ? eta -:--:--
   ---------------------------------------- 107.7/107.7 kB 1.2 MB/s eta 0:00:00
Using cached typing_extensions-4.15.0-py3-none-any.whl (44 kB)
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
pip install lxml

   ---------------------------------------- 0.0/4.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/4.0 MB ? eta -:--:--
    --------------------------------------- 0.1/4.0 MB 1.1 MB/s eta 0:00:04
   - -------------------------------------- 0.2/4.0 MB 1.5 MB/s eta 0:00:03
   -- ------------------------------------- 0.3/4.0 MB 1.9 MB/s eta 0:00:02
   --- ------------------------------------ 0.4/4.0 MB 1.9 MB/s eta 0:00:02
   ---- ----------------------------------- 0.5/4.0 MB 2.1 MB/s eta 0:00:02
   ------ --------------------------------- 0.6/4.0 MB 2.2 MB/s eta 0:00:02
   ------- -------------------------------- 0.7/4.0 MB 2.3 MB/s eta 0:00:02
   -------- ------------------------------- 0.8/4.0 MB 2.5 MB/s eta 0:00:02
   -------- ------------------------------- 0.9/4.0 MB 2.3 MB/s eta 0:00:02
   ---------- ----------------------------- 1.0/4.0 MB 2.4 MB/s eta 0:00:02
   ----------- ---------------------------- 1.2/4.0 MB 2.5 MB/s eta 0:00:02
   ------------ ----------


[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
# Membaca teks dari halaman website
import urllib.request
from bs4 import BeautifulSoup as bs

URL = "https://www.its.ac.id/informatika/fasilitas/laboratorium/laboratorium-komputasi-cerdas-dan-visi/"
doc = urllib.request.urlopen(URL).read()
doc = bs(doc, 'lxml').text

doc

"\n\n\n\n\nLaboratorium Komputasi Cerdas dan Visi - Departemen Teknik Informatika\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n \n\n\n\n\n\n\n\nToggle navigation\n\n\n\n\n\n\n\n\n\n\n\n\n\nCalon Mahasiswa\nMahasiswa\nDosen dan Staf\nAlumni\n\nid \n\nBahasa IndonesiaEnglish\n\n\n\n\r\n                            myITS \n\n \nTentang Informatika  \n\n\n\n\n\nSambutan Kepala Departemen\nVisi dan Misi\nSejarah\nPenghargaan\nPrestasi\nDepartemen Dalam Angka\nProgram Studi  \n\n\n\n\n\nTeknik Informatika (S1)\nTeknik Informatika (IUP)\nRekayasa Perangkat Lunak (S1)\nRekayasa Kecerdasan Artifisial (S1)\nTeknik Informatika (S2)\nIlmu Komputer (S3)\nPendaftaran\nFasilitas  \n\n\n\n\n\nLaboratorium\nRuang Aula\nRuang Kelas\nRuang Sidang\nRuang Baca\nFasilitas IT\nAkademik  \n\n\n\n\n\nPeraturan dan Informasi\nTugas Akhir Mahasiswa\nRiset dan Kolaborasi  \n\n\n\n\n\nBidang Penelitian\nPublikasi Ilmiah\n\n\n\n\n\n\n\

<h3>Membaca data dari modul</h3
<p>Contoh dataset dokumen yang cukup tenar: &quot;20 NewsGroup&quot;</p>

<img alt="" src="figures/6_20News.jpg" style="height: 300px ; width: 533px" />
<p><a href="http://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_20newsgroups.html#sklearn.datasets.fetch_20newsgroups" target="_blank">http://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_20newsgroups.html#sklearn.datasets.fetch_20newsgroups</a></p>

<p><strong>Categories </strong>=&nbsp;</p>
<pre>
[&#39;alt.atheism&#39;, &#39;comp.graphics&#39;, &#39;comp.os.ms-windows.misc&#39;, &#39;comp.sys.ibm.pc.hardware&#39;, &#39;comp.sys.mac.hardware&#39;,
 &#39;comp.windows.x&#39;, &#39;misc.forsale&#39;, &#39;rec.autos&#39;, &#39;rec.motorcycles&#39;, &#39;rec.sport.baseball&#39;, &#39;rec.sport.hockey&#39;,
 &#39;sci.crypt&#39;, &#39;sci.electronics&#39;, &#39;sci.med&#39;, &#39;sci.space&#39;, &#39;soc.religion.christian&#39;, &#39;talk.politics.guns&#39;,
 &#39;talk.politics.mideast&#39;, &#39;talk.politics.misc&#39;, &#39;talk.religion.misc&#39;]</pre>


In [2]:
pip install scikit-learn

  Using cached scikit_learn-1.8.0-cp312-cp312-win_amd64.whl.metadata (11 kB)
  Using cached scipy-1.17.1-cp312-cp312-win_amd64.whl.metadata (60 kB)
  Using cached joblib-1.5.3-py3-none-any.whl.metadata (5.5 kB)
  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
Using cached scikit_learn-1.8.0-cp312-cp312-win_amd64.whl (8.0 MB)
Using cached joblib-1.5.3-py3-none-any.whl (309 kB)
   ---------------------------------------- 0.0/12.3 MB ? eta -:--:--
   ---------------------------------------- 0.0/12.3 MB ? eta -:--:--
   ---------------------------------------- 0.0/12.3 MB 487.6 kB/s eta 0:00:26
   ---------------------------------------- 0.1/12.3 MB 1.4 MB/s eta 0:00:09
    --------------------------------------- 0.2/12.3 MB 1.5 MB/s eta 0:00:08
    --------------------------------------- 0.3/12.3 MB 1.9 MB/s eta 0:00:07
   - -------------------------------------- 0.4/12.3 MB 1.8 MB/s eta 0:00:07
   - -------------------------------------- 0.5/12.3 MB 1.8 MB/s eta 0:00:


[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
# Membaca teks dari modul
# note: butuh koneksi internet

from sklearn.datasets import fetch_20newsgroups

categories = ['sci.med', 'talk.politics.misc',  'rec.autos']
data = fetch_20newsgroups(subset='train', categories=categories,remove=('headers', 'footers', 'quotes'))

data.data[0]

'I want to get a car alarm and I am thinking about getting an Ungo Box.\n    Does anyone out there have any knowledge or experience with any of\n    these alarms?  How about price ranges for the different models?\n    Are these good car alarms?  Please email me any responces.\n\n                cak3@ns3.lehigh.edu'

In [3]:
# Agar tidak perlu mendownload kembali data, ada baiknya disimpan
# jika dibutuhkan tinggal membacanya dari drive
# data dapat disimpan ke dalam bentuk pickle
import pickle
import os

dst_name = "20newsgroups.pickle"
path = "data"
dst_path = os.path.join(path, dst_name)

# wb = write binary
# w = write
# rb = read binary
# r = read
with open(dst_path, 'wb') as fout:
    pickle.dump(data, fout)

In [4]:
# Membaca data dari pickle

src_name = dst_name
path = "data"
src_path = os.path.join(path, src_name)

with open(src_path, 'rb') as fin:
    data_pckl = pickle.load(fin)

data_pckl.data[0]

'I want to get a car alarm and I am thinking about getting an Ungo Box.\n    Does anyone out there have any knowledge or experience with any of\n    these alarms?  How about price ranges for the different models?\n    Are these good car alarms?  Please email me any responces.\n\n                cak3@ns3.lehigh.edu'

In [5]:
len(data_pckl.data)

1653

### Representasi vektor

<ul>
    <li>Data teks tidak dapat secara langsung dianalisis dengan model statistik/ data mining</li>
    <li>Data teks terlebih dahulu harus diolah melalui tahapan pre-processing dan kemudian direpresentasikan ke dalam bentuk yang lebih sederhana (vektor) agar dapat dengan mudah untuk diolah</li>
    <li>Akan tetapi perubahan data ke dalam bentuk yang lebih sederhana dapat berakibat pada hilangnya beberapa informasi dari data tersebut. Contohnya pada metode <strong> bag-of-words </strong> informasi urutan antar kata hilang </li>
</ul>

<h3> Vector Space Model (VSM) </h3>
<p> VSM merepresentasikan sebuah kata ke dalam bentuk vektor dalam suatu dimensi spasial. </p>

<p><img alt="" src="figures/vsm.png" style="width: 300px; height: 213px;" /></p>


<h4> Contoh representasi vektor antara term dengan dokumen: </h4>
<p><img alt="" src="figures/3_Bentuk umum representasi dokumen.JPG" style="height: 294px ; width: 620px" /></p>

<p><img alt="" src="figures/vsm_matrix.png" style="width: 500px; height: 283px;" /></p>

In [7]:
# Contoh pembentukan representasi vektor dengan perhitungan frekuensi (VSM-TF)
# dapat menggunakan library CountVectorizer pada sklearn
from sklearn.feature_extraction.text import CountVectorizer
corpus = [
'This is the first document.',
'This document is the second document.',
'And this is the third one.',
'Is this the first document?',
]
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(corpus)
print(vectorizer.get_feature_names_out())
print(X.toarray())

['and' 'document' 'first' 'is' 'one' 'second' 'the' 'third' 'this']
[[0 1 1 1 0 0 1 0 1]
 [0 2 0 1 0 1 1 0 1]
 [1 0 0 1 1 0 1 1 1]
 [0 1 1 1 0 0 1 0 1]]


<h4> TF.IDF-Term Weighting </h4>
<ul>
    <li> Pada representasi vektor sebelumnya nilai antar term dengan dokumen hanya berupa frekuensi kemunculan saja </li>
    <li> Nilai antar term dengan dokumen dapat digantikan dengan bobot TF.IDF </li>
    <li> TF (Term Frequency) menghitung frekuensi dari suatu kata pada dokumen; sementara IDF (Inverse Document Frequency) menghitung nilai informatif dari suatu kata berdasarkan tingkat kemunculannya antar dokumen </li>

</ul>

<p><img alt="" src="figures/3_tfidf logic.jpg" style="height:359px; width:638px" /></p>

<h4> Rumus perhitungan TFIDF </h4>
<p><img alt="" src="figures/3_rumus tfidf.png" style="height:370px; width:367px" /></p>

<p><img alt="" src="figures/3_variant tfidf.png" style="height:300px; width:955px" /></p>

In [9]:
# Contoh pembentukan representasi vektor dengan TFIDF (VSM-TFIDF)
from sklearn.feature_extraction.text import TfidfVectorizer
corpus = [
'This is the first document.',
'This document is the second document.',
'And this is the third one.',
'Is this the first document?',
]
vectorizer_tfidf = TfidfVectorizer()
X_tfidf = vectorizer_tfidf.fit_transform(corpus)
print(vectorizer_tfidf.get_feature_names_out())
print(X_tfidf.toarray())

['and' 'document' 'first' 'is' 'one' 'second' 'the' 'third' 'this']
[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]
 [0.         0.6876236  0.         0.28108867 0.         0.53864762
  0.28108867 0.         0.28108867]
 [0.51184851 0.         0.         0.26710379 0.51184851 0.
  0.26710379 0.51184851 0.26710379]
 [0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


In [11]:
# Contoh pembentukan representasi vektor dengan binary 0 atau 1 (VSM-binary)
from sklearn.feature_extraction.text import CountVectorizer
corpus = [
'This is the first document.',
'This document is the second document.',
'And this is the third one.',
'Is this the first document?',
]
vectorizer_binary = CountVectorizer(binary=True)
X_binary = vectorizer_binary.fit_transform(corpus)
print(vectorizer_binary.get_feature_names_out())
print(X_binary.toarray())

['and' 'document' 'first' 'is' 'one' 'second' 'the' 'third' 'this']
[[0 1 1 1 0 0 1 0 1]
 [0 1 0 1 0 1 1 0 1]
 [1 0 0 1 1 0 1 1 1]
 [0 1 1 1 0 0 1 0 1]]


<h3 id="Latihan:"><font color="blue">Latihan 1:</font></h3>

<p> Ubahlah data dari 20newsgroup dataset yang telah disimpan dalam bentuk pickle ke dalam representasi vektornya
<ul>
    <li> VSM binary </li>
    <li> VSM TF </li>
    <li> VSM TFIDF </li>
</ul>

<p> Apakah ukuran matriks yang terbentuk dari ketiga VSM tersebut sama? </p>
<p> Apakah baris kosong berpengaruh terhadap ukuran matriks yang terbentuk? </p>

In [12]:
!pip install gensim

  Using cached gensim-4.4.0-cp312-cp312-win_amd64.whl.metadata (8.6 kB)
  Using cached smart_open-7.5.1-py3-none-any.whl.metadata (24 kB)
Using cached gensim-4.4.0-cp312-cp312-win_amd64.whl (24.4 MB)
Using cached smart_open-7.5.1-py3-none-any.whl (64 kB)
   ---------------------------------------- 0.0/60.4 kB ? eta -:--:--
   ------ --------------------------------- 10.2/60.4 kB ? eta -:--:--
   -------------------- ------------------- 30.7/60.4 kB 435.7 kB/s eta 0:00:01
   ---------------------------------------- 60.4/60.4 kB 643.6 kB/s eta 0:00:00



[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [14]:
!pip install pandas

  Using cached pandas-3.0.1-cp312-cp312-win_amd64.whl.metadata (19 kB)
  Using cached tzdata-2025.3-py2.py3-none-any.whl.metadata (1.4 kB)
Using cached pandas-3.0.1-cp312-cp312-win_amd64.whl (9.7 MB)
Using cached tzdata-2025.3-py2.py3-none-any.whl (348 kB)



[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
import pickle
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [4]:
PICKLE_PATH = "data/20newsgroups.pickle"

with open(PICKLE_PATH, "rb") as f:
    obj = pickle.load(f)

In [6]:
if hasattr(obj, "data"):
    corpus = obj.data
elif isinstance(obj, dict):
    possible_keys = ["data", "texts", "documents", "corpus"]
    found = None
    for k in possible_keys:
        if k in obj:
            found = k
            break
    if found is None:
        raise ValueError(
            f"Pickle berbentuk dict, tapi key corpus tidak ditemukan. "
            f"Available keys: {list(obj.keys())}"
        )
    corpus = obj[found]
elif isinstance(obj, (list, tuple, np.ndarray, pd.Series)):
    corpus = obj
else:
    raise ValueError(f"Tipe objek pickle tidak dikenali: {type(obj)}")

corpus = ["" if x is None else str(x) for x in corpus]

In [7]:
n_docs = len(corpus)
n_empty_raw = sum(1 for doc in corpus if doc.strip() == "")

print("Jumlah dokumen:", n_docs)
print("Jumlah dokumen kosong (raw):", n_empty_raw)
print()

Jumlah dokumen: 1653
Jumlah dokumen kosong (raw): 60



In [8]:
# VSM BINARY
vec_binary = CountVectorizer(binary=True)
X_binary = vec_binary.fit_transform(corpus)

# VSM TF
vec_tf = CountVectorizer()
X_tf = vec_tf.fit_transform(corpus)

# VSM TF-IDF
vec_tfidf = TfidfVectorizer()
X_tfidf = vec_tfidf.fit_transform(corpus)

In [12]:
print("Shape VSM Binary :", X_binary.shape)
print("Shape VSM TF     :", X_tf.shape)
print("Shape VSM TFIDF  :", X_tfidf.shape)
print()

same_shape = (X_binary.shape == X_tf.shape == X_tfidf.shape)
print(same_shape)

Shape VSM Binary : (1653, 22578)
Shape VSM TF     : (1653, 22578)
Shape VSM TFIDF  : (1653, 22578)

True


In [10]:
zero_rows_binary = np.sum(X_binary.getnnz(axis=1) == 0)
zero_rows_tf = np.sum(X_tf.getnnz(axis=1) == 0)
zero_rows_tfidf = np.sum(X_tfidf.getnnz(axis=1) == 0)

print()
print("Jumlah baris all-zero pada Binary :", zero_rows_binary)
print("Jumlah baris all-zero pada TF     :", zero_rows_tf)
print("Jumlah baris all-zero pada TFIDF  :", zero_rows_tfidf)

print()
print("Jumlah fitur Binary :", len(vec_binary.get_feature_names_out()))
print("Jumlah fitur TF     :", len(vec_tf.get_feature_names_out()))
print("Jumlah fitur TFIDF  :", len(vec_tfidf.get_feature_names_out()))

print()
print("Contoh 20 fitur pertama Binary:")
print(vec_binary.get_feature_names_out()[:20])


Jumlah baris all-zero pada Binary : 61
Jumlah baris all-zero pada TF     : 61
Jumlah baris all-zero pada TFIDF  : 61

Jumlah fitur Binary : 22578
Jumlah fitur TF     : 22578
Jumlah fitur TFIDF  : 22578

Contoh 20 fitur pertama Binary:
['00' '000' '0000' '00014' '000k' '000mi' '000miles' '0010' '0033'
 '004021809' '00500' '008561' '00am' '00pm' '01' '011' '0119' '01720'
 '01830' '02']


In [13]:
print("\n KESIMPULAN:")
if same_shape:
    print("1. Ukuran matriks ketiga VSM SAMA.")
else:
    print("1. Ukuran matriks ketiga VSM TIDAK sama.")

print(
    "2. Baris kosong tidak menambah jumlah kolom/fitur, "
    "tetapi tetap dihitung sebagai baris dokumen dan biasanya menjadi all-zero row."
)


 KESIMPULAN:
1. Ukuran matriks ketiga VSM SAMA.
2. Baris kosong tidak menambah jumlah kolom/fitur, tetapi tetap dihitung sebagai baris dokumen dan biasanya menjadi all-zero row.


#### Custom TFIDF
* Menurut http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html
* default formula tf-idf yang digunakan sk-learn adalah:
* $tfidf = tf * log(\frac{N}{df+1})$ ==> Smooth IDF
* namun kita merubahnya menjadi:
* $tfidf = tf * log(\frac{N}{df})$ ==> Non Smooth IDF
* $tfidf = (1+log(tf)) * log(\frac{N}{df+1})$ ==> sublinear_tf, Smooth IDF
* $tfidf = (1+log(tf)) * log(\frac{N}{df})$ ==> sublinear_tf, Non Smooth IDF

In [14]:
corpus = [
'This is the first document.',
'This document is the second document.',
'And this is the third one.',
'Is this the first document?',
]
tfidf_vectorizer = TfidfVectorizer(smooth_idf= False, sublinear_tf=True, lowercase=True)
tfidf = tfidf_vectorizer.fit_transform(corpus)
print(tfidf.shape) # Sama
print(tfidf[0].data) # Hanya data ini yg berubah
print(tfidf[0].indices) # Letak kolomnya = tfidf

(4, 9)
[0.3645444  0.3645444  0.3645444  0.61722732 0.46941728]
[8 3 6 2 1]


<h4 id="Zipf's-Law:--&quot;power-law&quot;-:">Zipf&#39;s Law: &quot;power law&quot; :</h4>

<img alt="" src="figures/zipf.png" style="width: 456px; height: 270px;" />
<ul>
	<li><a href="https://plus.maths.org/content/mystery-zipf" target="_blank">https://plus.maths.org/content/mystery-zipf</a></li>
	<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4176592/" target="_blank">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4176592/</a></li>
</ul>

In [15]:
# frequency filtering di VSM
# Menghilangkan term yg memiliki frequency kemunculan di banyak dokumen > threshold
# dan term yang memiliki frequency kemunculan antar dokumen < threshold

tfidf_vectorizer = TfidfVectorizer()
tfidf_1 = tfidf_vectorizer.fit_transform(corpus)

tfidf_vectorizer_filter = TfidfVectorizer(max_df=0.75, min_df=1)
tfidf_2 = tfidf_vectorizer_filter.fit_transform(corpus)

print(tfidf_1.shape)
print(tfidf_2.shape)

(4, 9)
(4, 6)


<h3 id="Latihan:"><font color="blue">Latihan 2:</font></h3>

<ul>
    <li> Lakukan filtering pada VSM dari 20newsgroup dataset </li>
    <li> Diskusikanlah, mengapa menghilangkan term yang memiliki frekuensi kemunculan tinggi antar dokumen?</li>
    <li> Mengapa menghilangkan term yang memiliki frekuensi kemunculan rendah antar dokumen? </li>
    <li> Mengapa perlu melakukan filtering berdasarkan frekuensi? </li>
</ul>

In [16]:
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [ ]:
# VSM awal tanpa filtering
cv_raw = CountVectorizer()
X_raw = cv_raw.fit_transform(corpus)

# VSM hasil filtering
# min_df=2  -> buang term yang muncul di < 2 dokumen
# max_df=0.8 -> buang term yang muncul di > 80% dokumen
cv_filtered = CountVectorizer(min_df=2, max_df=0.8)
X_filtered = cv_filtered.fit_transform(corpus)

# TF-IDF hasil filtering
tfidf_filtered = TfidfVectorizer(min_df=2, max_df=0.8)
X_tfidf_filtered = tfidf_filtered.fit_transform(corpus)

print("--- SHAPE SEBELUM DAN SESUDAH FILTERING ---")
print("CountVectorizer tanpa filtering :", X_raw.shape)
print("CountVectorizer dengan filtering:", X_filtered.shape)
print("TFIDF dengan filtering          :", X_tfidf_filtered.shape)

--- SHAPE SEBELUM DAN SESUDAH FILTERING ---
CountVectorizer tanpa filtering : (4, 9)
CountVectorizer dengan filtering: (4, 2)
TFIDF dengan filtering          : (4, 2)


In [ ]:
raw_terms = np.array(cv_raw.get_feature_names_out())
filtered_terms = set(cv_filtered.get_feature_names_out())

df_raw = np.asarray((X_raw > 0).sum(axis=0)).ravel()

too_common_mask = df_raw > (0.8 * len(corpus))
too_rare_mask = df_raw < 2

too_common_terms = raw_terms[too_common_mask]
too_rare_terms = raw_terms[too_rare_mask]

removed_terms = [t for t in raw_terms if t not in filtered_terms]

In [20]:
print("\n--- RINGKASAN FILTERING ---")
print("Jumlah dokumen                    :", len(corpus))
print("Jumlah term sebelum filtering     :", len(raw_terms))
print("Jumlah term sesudah filtering     :", len(filtered_terms))
print("Jumlah term dibuang               :", len(removed_terms))
print("Jumlah term terlalu sering muncul :", len(too_common_terms))
print("Jumlah term terlalu jarang muncul :", len(too_rare_terms))

print("\nContoh term terlalu sering muncul:")
print(too_common_terms[:20])

print("\nContoh term terlalu jarang muncul:")
print(too_rare_terms[:20])

print("\nContoh term yang tersisa setelah filtering:")
print(sorted(list(filtered_terms))[:30])


--- RINGKASAN FILTERING ---
Jumlah dokumen                    : 4
Jumlah term sebelum filtering     : 9
Jumlah term sesudah filtering     : 2
Jumlah term dibuang               : 7
Jumlah term terlalu sering muncul : 3
Jumlah term terlalu jarang muncul : 4

Contoh term terlalu sering muncul:
['is' 'the' 'this']

Contoh term terlalu jarang muncul:
['and' 'one' 'second' 'third']

Contoh term yang tersisa setelah filtering:
['document', 'first']


Jawaban:

1. Menghilangkan term dengan frekuensi kemunculan tinggi antar dokumen perlu dilakukan
   karena term seperti ini biasanya kurang diskriminatif. Term tersebut muncul di terlalu
   banyak dokumen sehingga kurang membantu membedakan topik satu dokumen dari dokumen lain.

2. Menghilangkan term dengan frekuensi kemunculan rendah antar dokumen juga penting
   karena term yang sangat jarang sering bersifat noise, typo, atau terlalu spesifik,
   sehingga kontribusinya ke representasi umum corpus kecil tetapi menambah dimensi fitur.

3. Filtering berdasarkan frekuensi perlu dilakukan agar ukuran vocabulary lebih ringkas,
   representasi lebih bersih, dan proses komputasi lebih efisien.

In [ ]:
# Model VSM dapat dipergunakan untuk N-gram
# Bermanfaat untuk menangkap term yang berupa frase
tfidf_vectorizer_ngram = TfidfVectorizer(ngram_range = (1,2))
tfidf_ngram = tfidf_vectorizer_ngram.fit_transform(corpus)

print("=====Without N-gram=====")
print(tfidf_vectorizer.vocabulary_)
print("=====With N-gram=====")
print(tfidf_vectorizer_ngram.vocabulary_)

=====Without N-gram=====
{'this': 8, 'is': 3, 'the': 6, 'first': 2, 'document': 1, 'second': 5, 'and': 0, 'third': 7, 'one': 4}
=====With N-gram=====
{'this': 18, 'is': 6, 'the': 12, 'first': 4, 'document': 2, 'this is': 20, 'is the': 7, 'the first': 13, 'first document': 5, 'second': 10, 'this document': 19, 'document is': 3, 'the second': 14, 'second document': 11, 'and': 0, 'third': 16, 'one': 9, 'and this': 1, 'the third': 15, 'third one': 17, 'is this': 8, 'this the': 21}


<h3 id="Latihan:"><font color="blue">Latihan 3:</font></h3>

<ul>
    <li> Diskusikanlah, apakah n-gram efektif untuk menangkap term yang berupa frase? apakah hasil yang diperoleh dari n-gram SVM sudah memuaskan?
</ul>

In [ ]:
tfidf_uni = TfidfVectorizer(min_df=2, max_df=0.8, ngram_range=(1,1))
X_uni = tfidf_uni.fit_transform(corpus)

tfidf_ngram = TfidfVectorizer(min_df=2, max_df=0.8, ngram_range=(1,2))
X_ngram = tfidf_ngram.fit_transform(corpus)

uni_terms = tfidf_uni.get_feature_names_out()
ngram_terms = tfidf_ngram.get_feature_names_out()

bigram_terms = [t for t in ngram_terms if len(t.split()) == 2]

print("--- PERBANDINGAN UNIGRAM VS N-GRAM ---")
print("Shape tanpa n-gram :", X_uni.shape)
print("Shape dengan n-gram:", X_ngram.shape)

print("\nJumlah fitur unigram :", len(uni_terms))
print("Jumlah fitur n-gram  :", len(ngram_terms))
print("Jumlah bigram        :", len(bigram_terms))

print("\nContoh 30 bigram pertama:")
print(bigram_terms[:30])

--- PERBANDINGAN UNIGRAM VS N-GRAM ---
Shape tanpa n-gram : (4, 2)
Shape dengan n-gram: (4, 6)

Jumlah fitur unigram : 2
Jumlah fitur n-gram  : 6
Jumlah bigram        : 4

Contoh 30 bigram pertama:
['first document', 'is the', 'the first', 'this is']


In [25]:
import pickle

with open("data/20newsgroups.pickle", "rb") as f:
    data_obj = pickle.load(f)

print(type(data_obj))

if hasattr(data_obj, "keys"):
    print("Keys:", data_obj.keys())
else:
    print("Attributes:", dir(data_obj))

<class 'sklearn.utils._bunch.Bunch'>
Keys: dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])


In [29]:
from sklearn.pipeline import make_pipeline
from sklearn.svm import LinearSVC
from sklearn.model_selection import cross_val_score
import numpy as np

data_obj = obj

corpus = data_obj.data
y = data_obj.target

print(type(y))
print(y[:10])
print(len(corpus), len(y))

svm_uni = make_pipeline(
    TfidfVectorizer(min_df=2, max_df=0.8, ngram_range=(1,1)),
    LinearSVC()
)

svm_ngram = make_pipeline(
    TfidfVectorizer(min_df=2, max_df=0.8, ngram_range=(1,2)),
    LinearSVC()
)

score_uni = cross_val_score(svm_uni, corpus, y, cv=5, scoring="accuracy")
score_ngram = cross_val_score(svm_ngram, corpus, y, cv=5, scoring="accuracy")

print("Akurasi CV unigram      :", np.round(score_uni, 4))
print("Mean unigram            :", score_uni.mean())

print("\nAkurasi CV unigram+bigram:", np.round(score_ngram, 4))
print("Mean unigram+bigram      :", score_ngram.mean())

<class 'numpy.ndarray'>
[0 2 0 1 2 0 2 0 2 1]
1653 1653
Akurasi CV unigram      : [0.9215 0.8882 0.8671 0.9152 0.8879]
Mean unigram            : 0.8959534926302297

Akurasi CV unigram+bigram: [0.9124 0.8852 0.8671 0.9242 0.897 ]
Mean unigram+bigram      : 0.8971729378375904


Jawaban:

1. N-gram efektif untuk menangkap term yang berupa frase karena model tidak hanya melihat kata tunggal, tetapi juga pasangan kata yang muncul berurutan. Pada percobaan ini, model unigram menghasilkan mean accuracy sebesar 0.89595, sedangkan model unigram+bigram menghasilkan mean accuracy sebesar 0.89717.

2. Hasil tersebut menunjukkan bahwa penggunaan n-gram memberikan peningkatan performa, tetapi peningkatannya sangat kecil, yaitu sekitar 0.00122. Artinya, bigram membantu menangkap konteks frase, namun kontribusinya pada dataset ini belum terlalu signifikan.

3. Jadi, dapat disimpulkan bahwa n-gram memang efektif untuk merepresentasikan frase, tetapi hasil n-gram SVM pada percobaan ini belum jauh lebih baik dibandingkan unigram saja. Hasilnya sudah cukup baik karena performanya sedikit meningkat, tetapi belum bisa dikatakan memberikan perbedaan yang besar.